menpofit/dlib/fitter.py

```diff
@@ -0,0 +1,276 @@
+
+            # Train the Dlib model
+            current_bounding_boxes = self.algorithms[j].train(
+                scaled_images, scaled_gt_shapes, current_bounding_boxes,
+                prefix=scale_prefix, verbose=verbose)
+
+            # Scale current shapes to next resolution, don't bother
+            # scaling final level
+            if j != (self.n_scales - 1):
+                transform = Scale(self.scales[j + 1] / self.scales[j],
+                                  n_dims=2)
+                for bboxes in current_bounding_boxes:
+                    for bb in bboxes:
+                        transform.apply_inplace(bb)
+
+    @property
+    def n_scales(self):
+        """
+        The number of scales of the Dlib fitter.
+
+        :type: `int`
+        """
+        return len(self.scales)
+
+    def perturb_from_bb(self, gt_shape, bb):
+        return self._perturb_from_gt_bounding_box(gt_shape, bb)
+
+    def perturb_from_gt_bb(self, gt_bb):
+            return self._perturb_from_gt_bounding_box(gt_bb, gt_bb)
+
+    def _fitter_result(self, image, algorithm_results, affine_correction,
+                       gt_shape=None):
+        return MultiFitterResult(image, self, algorithm_results,
+                                 affine_correction, gt_shape=gt_shape)
+
+    def fit_from_shape(self, image, initial_shape, max_iters=20, gt_shape=None,
+                       crop_image=None, **kwargs):
+
+        warnings.warn('Fitting from an initial shape is not supported by '
+                      'Dlib - therefore we are falling back to the tightest '
+                      'bounding box from the given initial_shape')
+        tightest_bb = initial_shape.bounding_box()
+        return self.fit_from_bb(image, tightest_bb, max_iters=max_iters,
+                                gt_shape=gt_shape, crop_image=crop_image,
+                                **kwargs)
+
+    def fit_from_bb(self, image, bounding_box, max_iters=20, gt_shape=None,
+                    crop_image=None, **kwargs):
+        # generate the list of images to be fitted
+        images, bounding_boxes, gt_shapes = self._prepare_image(
+            image, bounding_box, gt_shape=gt_shape, crop_image=crop_image)
+
+        # work out the affine transform between the initial shape of the
+        # highest pyramidal level and the initial shape of the original image
+        affine_correction = AlignmentAffine(bounding_boxes[-1], bounding_box)
+
+        # run multilevel fitting
+        algorithm_results = self._fit(images, bounding_boxes[0],
+                                      max_iters=max_iters,
+                                      gt_shapes=gt_shapes, **kwargs)
+
+        # build multilevel fitting result
+        fitter_result = self._fitter_result(
+            image, algorithm_results, affine_correction, gt_shape=gt_shape)
+
+        return fitter_result
+
+    def __str__(self):
+        return ''
```

Comment:
@patricksnape may return something here? You have good knowledge of the information stored in this fitter object, so just return something basic.